### ISSUE TITLE
Request for a GPUCommandEncoder fillBuffer operation

### ISSUE THREAD

**Issue by benvanik** [gpuweb/gpuweb#1968]
There's currently no way to fill a buffer from the device using a hardware-supported DMA (or driver-optimized) fill operation. A fill operation is available in practically all APIs ([MTLBlitCommandEncoder fillBuffer:range:value:](https://developer.apple.com/documentation/metal/mtlblitcommandencoder/1400761-fillbuffer?language=objc), [vkCmdFillBuffer](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/vkCmdFillBuffer.html), [ID3D12GraphicsCommandList::ClearUnorderedAccessViewUnit](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-id3d12graphicscommandlist-clearunorderedaccessviewuint) - [example](https://github.com/gfx-rs/gfx/blob/2c8db0b1322f79e1fcd56ead828ea20908769254/src/backend/dx12/src/command.rs#L2207-L2274), etc), and when not it'd be significantly more efficient for the WebGPU implementation to emulate it instead of user code restricted to the WebGPU API.

When emulating fill with a shader in user-mode the hardware is unable to take advantage of dedicated DMA resources that may be available and what should be a relatively cheap accelerated memset instead occupies compute units. It's also much more complex for user code to build new bind groups and issue multiple commands when performing the fill vs. a single GPUCommandEncoder fillBuffer that specified just the destination buffer, offset, size, and value inline. Due to the lack of push constants it also means that a uniform buffer update is required to shuffle the fill value to the shader which can introduce significant complexity when trying to emulate a fillBuffer in middleware as even more bind groups are required, etc. Even worse if each fill range has a different size (they often do) then either new unique bind groups are required that specify the size (more API overhead) or one must use a single bind group that specifies the whole buffer with the size passed through uniforms - which without explicit barriers means two fills to subranges of the same buffer (or a fill and a dispatch, etc) will have a false dependency.

It'd be great to see this addition to the API to make it easier to tunnel through to the native APIs that expose this functionality as well as more efficiently utilize hardware resources. AFAICS the same reason one has GPUCommandEncoder copyBufferToBuffer instead of emulating that with a shader applies here.

Proposal:
```
interface GPUCommandEncoder {
+ undefined fillBuffer(GPUBuffer buffer, GPUSize64 offset, GPUSize64 length, unsigned long value); 
}
```

**Comment by Kangz** [comment]
Note that FillBuffer, at least on AMD / Intel / Nvidia will be done using a compute shader (I checked the Intel and AMD Mesa drivers, and that's suggested by the D3D12 function as well). Mesa's Qualcomm driver uses a blit though, and this might also be the case on ARM (but that Vulkan function is still a stub in Mesa).

Overall this is a simple addition to the API and the biggest open question is what the usage of the buffer should be for the operation (Vulkan requires `COPY_DST` while D3D12 would require `STORAGE`). However it can be emulated by the application (even if that has a higher cost) so I don't think we should add it for the MVP / V1. IMHO it seems more useful to get WebGPU out faster than adding this method in the first version.

**Comment by benvanik** [comment]
Makes sense! I'd of course love to have it sooner - having worked with wasm I'm used to "post-MVP" meaning 5+yr and am hoping something simple like this doesn't stretch that far - but at least as you say it is possible to emulate. I'd put this in the bucket of "making the WebGPU API match the compute surface area of basic native APIs" that hopefully wouldn't be as involved a process as things like multi-queue understandably are :)

The downside to emulation is that hardware that does have dedicated transfer resources/fixed function support/optimized implementations can't be utilized and it's a much harder lift in user mode in terms of binary size, runtime cost of additional pipelines to compile, code to manage it all, API overhead for setting it all up per invocation vs. a single call, etc. In our uses we care more about the former - we have ~hundreds of shaders to compile so one more is not going to break us - but there's also a user convivence thing: a data manipulation API isn't complete without a memset (or, with the other GPUQueue.readBuffer request, memcpy). It's mostly just a bit unfortunate that new WebGPU users will need to implement memset themselves each time and polyfills/libraries are difficult given the other API design points so it almost always makes sense to go through the pain of rolling your own.

Once we have something working we can benchmark and compare a native driver-provided implementation vs the WebGPU emulated one and see what % of peak we reach. When we've tried this in the past (if you want to fill with a unique 64-bit value, for example, you have to do that yourself) it's not been great even when directly using CUDA and Vulkan. The numbers I still have at hand from the CUDA benchmark showed 303GB/s vs. 1029GB/s for emulated (launch of a multibyte memset kernel) vs. driver (cudaMemsetAsync). When you're working on memory-bound problems that kind of difference really matters :)

**Comment by litherum** [comment]
The Vulkan and Metal implementations are incompatible - Metal repeats a single byte over and over again, whereas Vulkan repeats a 4-byte pattern over and over again.

**Comment by litherum** [comment]
I guess we could limit it to just zeroing out a (range of a) buffer. That's mainly what it's used for in Metal, anyway.

**Comment by litherum** [comment]
I guess you could implement the Metal behavior on Vulkan (by building a 4-byte pattern that repeats the input byte 4 times) but you couldn't implement the Vulkan behavior on Metal.

**Comment by kainino0x** [comment]
There are some complexities here, especially that D3D12 requires a UAV in order to clear.

However our implementations already need to be able to clear buffers _to zero_ (for implicit initialization), which we can do by copying from pre-cleared memory. @benvanik how often do you need to fill to values other than 0? If rarely, maybe we could start with a simple version of this that only allows the clear value to be 0.

**Comment by kainino0x** [comment]
See also #28

**Comment by litherum** [comment]
> @benvanik how often do you need to fill to values other than 0?

(Assuming you probably will never be able to fill 32-bit values in WebGPU and would instead only be able to fill 8-bit values)

**Comment by kainino0x** [comment]
Resolution: we should have a way to zero out subranges of buffers. TBD on whether there are byte alignment restrictions on that.

@benvanik your feedback on whether this is sufficient would still be very helpful.

**Comment by benvanik** [comment]
TLDR: zeroing would be enough and very helpful!

In our particular cases we do have some that are non-zero (in the ML space a lot of initialization vectors are non-zero splats) - kind of like how in image processing you may want to set the A byte of an RGBA texture to 0xFF or if doing YUV the resetting to the 0x80 center point. However a bulk of the cases that happen frequently enough to matter are zeroing when reusing buffers in user-space via pooling/ringbuffers/etc zero'ing - emulating the fills for non-zero values initialized once isn't bad.

The particular advantage our approach has is that if we can shift the math around - so as long as we can zero something we can fill it to any value T by making the first access (read or write) be zero+T - we just need a way to get the reused buffers into predictable states.

**Comment by Kangz** [comment]
Deduplicating #2009

@litherum said:

> We might need to restrict the offset / length to be multiples of 4, because of Vulkan or D3D. We should investigate whether or not we need it.

We should keep this in mind when implementing the feature.
